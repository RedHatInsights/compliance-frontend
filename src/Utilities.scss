// Sass Imports

@import "~@redhat-cloud-services/frontend-components-utilities/styles/_all";
