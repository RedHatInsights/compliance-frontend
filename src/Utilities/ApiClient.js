class ApiClient {
    DEFAULT_API_HEADERS = {
        'Content-Type': 'application/json',
        Accept: 'application/json'
    }

    constructor(options = {}) {
        this.apiBase = options.apiBase || '/api';
        this.path = options.path || '';
        this.authenticate = options.authenticate || (() => ({}));
        this.onUnauthorised = options.onUnauthorised;
        this.defaultHeaders = options.headers || this.DEFAULT_API_HEADERS;
    }

    async request(path, apiProps, method, options = {}) {
        return await this.callAuthenticate()
        .then(() => this.fetch(path, apiProps, method))
        .then(this.checkForEmptyResponse)
        .then((response) => this.checkForErrors(response, options))
        .then((response) => {
            const json = (async () => await response.json());
            return json();
        })
        .catch(this.finalCatch);
    }

    finalCatch(promise) {
        if (promise.errors === undefined) {
            return Promise.reject({ title: 'Error parsing' });
        } else {
            return Promise.reject({ ...promise });
        }
    }

    fetch(path, apiProps, method) {
        let params = {
            method: method || 'get',
            headers: this.defaultHeaders,
            credentials: 'include'
        };

        if (apiProps) {
            params.body = JSON.stringify(apiProps);
        }

        return fetch(
            `${ this.apiBase }${ this.path ? this.path : '' }${ path ? path : '' }`, params
        );
    }

    checkForEmptyResponse(response) {
        return response.status === 204 ? { json: () => ({}) } : response;
    }

    checkForErrors(response, options = {}) {
        if (response.status === 404 && options.ignore404) {
            return { json: () => ({}) };
        }

        if (response.status === 401) {
            return this.onUnauthorised();
        }

        const responseCloneJson = response.clone ? response.clone().json() : response;

        if (response.status === 422) {
            return responseCloneJson.then((json) =>
                Promise.reject({ ...json, title: 'Validation error' })
            );
        }

        if (response.status >= 400 && response.status <= 600) {
            return responseCloneJson.then((json) =>
                Promise.reject(json.errors[0])
            );
        }

        return response;
    }

    callAuthenticate() {
        return Promise.resolve(this.authenticate() || true);
    }

    create(path, apiProps) {
        return this.request(path, apiProps, 'post');
    }

    update(path, apiProps) {
        return this.request(path, apiProps, 'put');
    }

    get(path, options = {}) {
        return this.request(path, null, 'get', options);
    }

    destroy(path) {
        return this.request(path, null, 'delete');
    }
}

export default ApiClient;
