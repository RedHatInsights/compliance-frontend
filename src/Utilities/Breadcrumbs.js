import { COMPLIANCE_UI_ROOT } from '../constants';

export function onNavigate(event, history) {
    event.preventDefault();
    const path = event.target.pathname.replace(COMPLIANCE_UI_ROOT, '');
    (history ? history : this.props.history).push(path);
}
