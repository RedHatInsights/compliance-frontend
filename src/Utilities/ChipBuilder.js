import { conditionalFilterType } from '@redhat-cloud-services/frontend-components/ConditionalFilter';

const findWithString = (value) => (
    (item) => (
        String(item.value) === String(value)
    )
);

class ChipBuilder {
    constructor(filterConfig) {
        this.filterConfig = filterConfig;
        this.config = this.filterConfig.config;
    }

    textFilterChip = (category, currentValue) => (currentValue && currentValue !== '' ? {
        category,
        chips: [{ name: currentValue }]
    } : null)

    checkboxFilterChip = (category, currentValue) => (currentValue && currentValue.length > 0 ? {
        category,
        chips: currentValue.map((value) => (
            { name: this.filterConfig.labelForValue(value, category) }
        ))
    } : null)

    radioFilterChip = (category, currentValue) => (currentValue && currentValue !== '' ? {
        category,
        chips: [{ name: this.filterConfig.labelForValue(currentValue, category) }]
    } : null)

    groupFilterChip = (category, currentValue = {}, categoryConfig) => {
        const items = Object.keys(currentValue).flatMap((groupValue) => {
            const group = categoryConfig.items.find(findWithString(groupValue));

            return Object.keys(currentValue[groupValue]).map((itemValue) => (
                currentValue[groupValue][itemValue] ? group?.items.find(findWithString(itemValue)) : null
            ));
        }).filter((v) => (!!v));

        return currentValue && items.length > 0 ? {
            category,
            chips: items.map((item) => ({ name: item.label }))
        } : null;
    }

    chipFor = (filter, currentValue) => {
        const categoryConfig = this.filterConfig.getCategoryForLabel(filter);
        const { label, type } = categoryConfig ? categoryConfig : { label: filter, type: null };

        switch (type) {
            case conditionalFilterType.text:
                return this.textFilterChip(label, currentValue);

            case conditionalFilterType.checkbox:
                return this.checkboxFilterChip(label, currentValue);

            case conditionalFilterType.radio:
                return this.radioFilterChip(label, currentValue);

            case conditionalFilterType.group:
                return this.groupFilterChip(label, currentValue, categoryConfig);

            default:
                return null;
        }
    }

    chipsFor = (filters) => (
        Object.keys(filters).map((filter) => (
            this.chipFor(filter, filters[filter])
        )).filter((f) => (!!f))
    )
}

export default ChipBuilder;
