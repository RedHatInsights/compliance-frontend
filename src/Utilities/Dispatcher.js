import { getStore } from 'Store';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/redux';

export function dispatchAction(actionCreator) {
  const store = getStore();
  return store.dispatch(actionCreator);
}

export const dispatchNotification = (notification) => {
  dispatchAction(addNotification(notification));
};
