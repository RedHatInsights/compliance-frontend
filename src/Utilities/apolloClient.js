import { ApolloClient, HttpLink, InMemoryCache } from 'apollo-boost';

import { COMPLIANCE_API_ROOT } from '@/constants';

const client = new ApolloClient({
  link: new HttpLink({
    credentials: 'include',
    uri: COMPLIANCE_API_ROOT + '/graphql',
  }),
  cache: new InMemoryCache(),
});

export default client;
