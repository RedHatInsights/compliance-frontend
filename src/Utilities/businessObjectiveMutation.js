import { CREATE_BUSINESS_OBJECTIVE } from 'Utilities/graphql/mutations';

const businessObjectiveMutation = (businessObjective, editPolicyBusinessObjective, mutate) => {
    if (editPolicyBusinessObjective === undefined) {
        return Promise.resolve(businessObjective ? businessObjective.id : null);
    }

    if (editPolicyBusinessObjective && !editPolicyBusinessObjective.create && businessObjective
        && (editPolicyBusinessObjective.value !== businessObjective.id)) {
        return Promise.resolve(editPolicyBusinessObjective.value);
    }

    if (editPolicyBusinessObjective && !editPolicyBusinessObjective.create
        && businessObjective === null) {
        return Promise.resolve(editPolicyBusinessObjective.value);
    }

    if (editPolicyBusinessObjective === null) {
        return Promise.resolve(null);
    }

    if (editPolicyBusinessObjective.create) {
        return mutate({
            mutation: CREATE_BUSINESS_OBJECTIVE,
            variables: { input: { title: editPolicyBusinessObjective.label } }
        }).then((result) => {
            return result.data.createBusinessObjective.businessObjective.id;
        });
    }

    return Promise.resolve();
};

export default businessObjectiveMutation;
