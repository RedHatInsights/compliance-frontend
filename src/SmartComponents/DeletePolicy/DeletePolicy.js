import {
    Modal,
    TextContent,
    Button
} from '@patternfly/react-core';
import propTypes from 'prop-types';
import React from 'react';
import { useMutation } from '@apollo/react-hooks';
import { DELETE_PROFILE } from '../../Utilities/graphql/mutations';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications';
import { dispatchAction } from '../../Utilities/Dispatcher';

const DeletePolicy = ({ isModalOpen, policy, toggle, onDelete }) => {
    const [deletePolicy] = useMutation(DELETE_PROFILE, {
        onCompleted: () => {
            dispatchAction(addNotification({
                variant: 'success',
                title: `Removed policy ${policy.name}`
            }));
            onDelete();
            toggle();
        },
        onError: (error) => {
            dispatchAction(addNotification({
                variant: 'danger',
                title: 'Error removing policy',
                description: error.message
            }));
            onDelete();
            toggle();
        }
    });
    const { name, id } = policy;

    return (
        <Modal
            isSmall
            title='Delete policy'
            isOpen={isModalOpen}
            isFooterLeftAligned
            onClose={toggle}
            actions={[
                <Button key='destroy'
                    aria-label="delete"
                    variant='danger'
                    onClick={() => deletePolicy({ variables: { input: { id } } })}>
                    Delete policy
                </Button>,
                <Button key='cancel' variant='secondary' onClick={toggle}>
                    Cancel
                </Button>
            ]}
        >
            <TextContent>
                Are you sure you want to delete <b>{ name }</b>?
                This cannot be undone.
            </TextContent>
        </Modal>
    );
};

DeletePolicy.propTypes = {
    policy: propTypes.object,
    toggle: propTypes.func,
    isModalOpen: propTypes.bool,
    onDelete: propTypes.func
};

export default DeletePolicy;
