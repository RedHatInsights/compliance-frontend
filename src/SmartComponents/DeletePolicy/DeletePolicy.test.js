import { useLocation } from 'react-router-dom';
import { useMutation } from '@apollo/client';
import { dispatchAction } from 'Utilities/Dispatcher';

import DeletePolicy from './DeletePolicy.js';

jest.mock('Utilities/Dispatcher');

jest.mock('@apollo/client', () => ({
    ...jest.requireActual('@apollo/client'),
    useMutation: jest.fn(() => ({}))
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: jest.fn(() => ({})),
    useHistory: jest.fn(() => ({
        push: jest.fn()
    }))
}));

describe('DeletePolicy', () => {
    const policy = { id: 1, name: 'foo' };

    beforeEach(() => {
        useMutation.mockImplementation((query, options) => {
            return [function() {
                options.onCompleted();
            }];
        });
        useLocation.mockImplementation(() => ({
            state: {
                policy
            }
        }));
        dispatchAction.mockImplementation(() => {});
    });

    it('expect to render an open modal without error', () => {
        const component = shallow(
            <DeletePolicy />
        );

        expect(toJson(component)).toMatchSnapshot();
    });
});
