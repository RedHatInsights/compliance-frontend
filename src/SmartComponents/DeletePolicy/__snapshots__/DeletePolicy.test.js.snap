// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeletePolicy expect to render an open modal without error 1`] = `
<Modal
  actions={
    Array [
      <Button
        aria-label="delete"
        isDisabled={true}
        onClick={[Function]}
        variant="danger"
      >
        Delete policy and associated reports
      </Button>,
      <Button
        onClick={[Function]}
        variant="secondary"
      >
        Cancel
      </Button>,
    ]
  }
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  showClose={true}
  title={
    <React.Fragment>
      <ExclamationTriangleIcon
        className="ins-u-warning"
        color="currentColor"
        noVerticalAlign={false}
        size="sm"
      />
      <Text
        className="policy-delete-header-text"
        component="span"
      >
        Delete policy?
      </Text>
    </React.Fragment>
  }
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <Text
    className="policy-delete-body-text"
  >
    Deleting the policy 
    <b>
      foo
    </b>
     will also delete its associated reports.
  </Text>
  <Checkbox
    className=""
    id="deleting-policy-check-1"
    isChecked={false}
    isDisabled={false}
    isValid={true}
    label="I understand this will delete the policy and all associated reports"
    onChange={[Function]}
  />
</Modal>
`;
